File: 백준/Gold/1916. 최소비용 구하기/최소비용 구하기.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.*;

class Pair implements Comparable<Pair>{
    int distance;
    int startPoint;

    public Pair(int distance, int startPoint) {
        this.distance = distance;
        this.startPoint = startPoint;
    }

    @Override
    public int compareTo(Pair other) {
        return Integer.compare(this.distance, other.distance);
    }
}

public class Main {
    static ArrayList<Pair>[] graph;
    static int[] minCost;
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st;

        int N = Integer.parseInt(br.readLine()); //도시의 개수
        int M = Integer.parseInt(br.readLine()); //버스의 개수

        graph = new ArrayList[N+1];
        minCost = new int[N+1];
        Arrays.fill(minCost, Integer.MAX_VALUE); //도시까지 가는 경로 Max로 초기화

        for(int i=0;i<graph.length;++i){
            graph[i] = new ArrayList();
        }

        for(int i=0;i<M;++i){
            st = new StringTokenizer(br.readLine());
            //출발지 인덱스에 도착지, 비용 저장

            int start = Integer.parseInt(st.nextToken());
            int end = Integer.parseInt(st.nextToken());
            int cost = Integer.parseInt(st.nextToken());

            if(start == end)
                continue;

            graph[start].add(new Pair(cost, end));
        }

        st = new StringTokenizer(br.readLine());
        int start = Integer.parseInt(st.nextToken());
        int end = Integer.parseInt(st.nextToken());

        dijkstra(start);

        System.out.println(minCost[end]);
    }

    private static void dijkstra(int start) {
        PriorityQueue<Pair> q = new PriorityQueue<>();
        q.offer(new Pair(0, start));

        while (!q.isEmpty()){
            Pair poll = q.poll();

            if (minCost[poll.startPoint] < poll.distance) //지금까지의 최소거리가 아니면 무시하기
                continue;

            for(Pair next : graph[poll.startPoint]){ //현재 지점과 연결된 간선 정보 불러오기
                int sum = poll.distance + next.distance; //현재 지점까지의 비용 + 다음 도시까지의 비용
                if(minCost[next.startPoint] > sum){ //비용 갱신
                    minCost[next.startPoint] = sum;
                    q.offer(new Pair(sum, next.startPoint));
                }
            }
        }
    }
}

File: 백준/Gold/1916. 최소비용 구하기/최소비용 구하기.py
import heapq
import sys
input = sys.stdin.readline

INF = float('inf')

N = int(input()) #도시의 개수
M = int(input()) #버스의 개수

graph = [[] for _ in range(N+1)]

#출발지, 목적지, 비용 입력
for _ in range(M):
    values = input().split()

    start = int(values[0])
    end = int(values[1])
    distance = int(values[2])

    graph[start].append((end, distance))


result = [INF for _ in range(N+1)] #각 도시까지 가는 겨올 MAX로 초기화
wantS, wantE = map(int, input().split())

def dijkstra(start):
    q = []
    heapq.heappush(q, (0, start)) #해당 노드까지의 거리, 시작노드
    #result[start] = 0

    while q:
        dis, node = heapq.heappop(q)

        if result[node] < dis: #지금 저장되어 있는 최소거리보다 길다면, 예전 원소이므로 무시
            continue

        for next in graph[node]: #해당 노드와 인접 노드 탐색
            cost = dis + next[1] #지금까지의 거리 + 인접 노드까지의 거리
            if result[next[0]] > cost: #인접 노드까지의 거리 갱신
                result[next[0]] = cost
                heapq.heappush(q, (cost, next[0]))

dijkstra(wantS)
print(result[wantE]) #목적지까지의 최소 cost 구하기